armadillo/src/main/java/org/molgenis/armadillo/controller/StorageController.java

```diff
@@ -0,0 +1,292 @@
+        auditor.audit(
+            () -> storage.hasProject(project), principal, GET_PROJECT, Map.of(PROJECT, project));
+    return projectExists ? noContent().build() : notFound().build();
+  }
+
+  @Operation(summary = "List objects in a project")
+  @ApiResponses(
+      value = {
+        @ApiResponse(
+            responseCode = "200",
+            description = "Objects listed",
+            content =
+                @Content(array = @ArraySchema(schema = @Schema(implementation = String.class)))),
+        @ApiResponse(
+            responseCode = "404",
+            description = "Project does not exist",
+            content = @Content(schema = @Schema(hidden = true))),
+        @ApiResponse(
+            responseCode = "401",
+            description = "Unauthorized",
+            content = @Content(schema = @Schema(hidden = true)))
+      })
+  @GetMapping(
+      value = "/projects/{project}/objects",
+      produces = {APPLICATION_JSON_VALUE})
+  @ResponseStatus(OK)
+  public List<String> listObjects(Principal principal, @PathVariable String project) {
+    return auditor.audit(
+        () -> storage.listObjects(project), principal, LIST_OBJECTS, Map.of(PROJECT, project));
+  }
+
+  @Operation(summary = "Upload an object to a project")
+  @ApiResponses(
+      value = {
+        @ApiResponse(responseCode = "204", description = "Object uploaded successfully"),
+        @ApiResponse(responseCode = "404", description = "Unknown project"),
+        @ApiResponse(responseCode = "409", description = "Object already exists"),
+        @ApiResponse(responseCode = "401", description = "Unauthorized")
+      })
+  @PostMapping(
+      value = "/projects/{project}/objects",
+      consumes = {MULTIPART_FORM_DATA_VALUE})
+  @ResponseStatus(NO_CONTENT)
+  public void uploadObject(
+      Principal principal,
+      @PathVariable String project,
+      @RequestParam String object,
+      @RequestParam MultipartFile file) {
+    auditor.audit(
+        () -> addObject(project, object, file),
+        principal,
+        UPLOAD_OBJECT,
+        Map.of(PROJECT, project, OBJECT, object));
+  }
+
+  private void addObject(String project, String object, MultipartFile file) {
+    try {
+      storage.addObject(project, object, file.getInputStream());
+    } catch (IOException e) {
+      throw new FileProcessingException();
+    }
+  }
+
+  @Operation(summary = "Copy an object within a project")
+  @ApiResponses(
+      value = {
+        @ApiResponse(responseCode = "204", description = "Object copied successfully"),
+        @ApiResponse(responseCode = "404", description = "Unknown project or object"),
+        @ApiResponse(responseCode = "409", description = "Object already exists"),
+        @ApiResponse(responseCode = "401", description = "Unauthorized")
+      })
+  @PostMapping(
+      value = "/projects/{project}/objects/{object}/copy",
+      consumes = {APPLICATION_JSON_VALUE})
+  @ResponseStatus(NO_CONTENT)
+  public void copyObject(
```

Comment:
I must confess this operation confused me. I do understand we then need to post a new name for the rename but it feels a bit clunky (aggrevated by swagger, where I simply expected an 'oldName' and 'newName' parameter instead of the third parameter being a json object.

Moreover, I can imagine users would like to copy or move to another project? So we might want to expand requestbody to include project as optional?

Reply:
I agree, this one isn't the easies to use (in Swagger) :/ It's because of the mix of path values and request body. I don't really see a better way, do you?

Good point about the copying to another project. Do you want to do it now? We can also wait until it's requested because adding an optional request body paramater will be backwards compatible.

Reply:
Hmmm, how about

`/storage/projects/{project}/objects/{object}/copyTo/{object}`

this wouldn't work for copying accross projects though, but would simplify usage (in Swagger) greatly